File: oop/kotlin/src/ConnectFour.kt
// ConnectFour with Move = Int (column), Undo = ConnectFourUndo.
//
// Parameterized on (rows, cols, need) so the same code covers the
// standard 6x7-need-4 game and small variants useful for benchmarking
// optimal play (e.g. 4x4-need-3). The full 6x7 game is far beyond a
// naive minimax — it's left as a hook for richer search strategies.

package games

data class ConnectFourUndo(val col: Int, val row: Int, val prevWinner: Side?)

class ConnectFour(
    private val rows: Int = 6,
    private val cols: Int = 7,
    private val need: Int = 4,
) : Game<Int, ConnectFourUndo> {

    private val cells: Array<CharArray> = Array(rows) { CharArray(cols) { '.' } }
    private var current: Side = Side.ONE
    private var _winner: Side? = null

    private fun mark(s: Side): Char = if (s == Side.ONE) 'X' else 'O'

    // Count consecutive marks in (dr,dc) from (r,c), exclusive.
    private fun run(r: Int, c: Int, dr: Int, dc: Int, m: Char): Int {
        var n = 0
        for (i in 1 until need) {
            val rr = r + dr * i
            val cc = c + dc * i
            if (rr !in 0 until rows || cc !in 0 until cols) break
            if (cells[rr][cc] != m) break
            n++
        }
        return n
    }

    private fun checkWinAt(r: Int, c: Int): Boolean {
        val m = cells[r][c]
        val dirs = arrayOf(0 to 1, 1 to 0, 1 to 1, 1 to -1)
        return dirs.any { (dr, dc) ->
            1 + run(r, c, dr, dc, m) + run(r, c, -dr, -dc, m) >= need
        }
    }

    override fun legalMoves(): List<Int> {
        if (_winner != null) return emptyList()
        return (0 until cols).filter { cells[0][it] == '.' }
    }

    override fun applyMove(move: Int): ConnectFourUndo {
        val prevWinner = _winner
        var row = rows - 1
        while (row >= 0 && cells[row][move] != '.') row--
        cells[row][move] = mark(current)
        if (checkWinAt(row, move)) _winner = current
        current = current.other()
        return ConnectFourUndo(move, row, prevWinner)
    }

    override fun undoMove(undo: ConnectFourUndo) {
        cells[undo.row][undo.col] = '.'
        _winner = undo.prevWinner
        current = current.other()
    }

    override fun currentSide(): Side = current
    override fun isOver(): Boolean = _winner != null || legalMoves().isEmpty()
    override fun winner(): Side? = _winner

    override fun stateKey(): Long {
        // rows*cols cells * 2 bits + 1 side bit. Empty=0, X=1, O=2.
        // Caller is responsible for using a small enough board to fit u64
        // (i.e. rows*cols <= 31). Benchmark uses 4x4 = 16 cells -> 33 bits.
        var k = 0L
        var i = 0
        for (r in 0 until rows) {
            for (c in 0 until cols) {
                val v = when (cells[r][c]) { '.' -> 0L; 'X' -> 1L; else -> 2L }
                k = k or (v shl (2 * i))
                i++
            }
        }
        k = k or ((if (current == Side.ONE) 0L else 1L) shl (2 * i))
        return k
    }

    override fun render(): String = buildString {
        for (r in 0 until rows) {
            append('|')
            for (c in 0 until cols) { append(cells[r][c]); append('|') }
            append('\n')
        }
        append(' ')
        for (c in 0 until cols) { append(c); append(' ') }
        append('\n')
    }

    override fun moveToString(move: Int): String = "col $move"
}

File: oop/rust/src/connectfour.rs
// ConnectFour with Move = usize (column index), Undo = ConnectFourUndo.
//
// Parameterized on (rows, cols, need) so the same code covers the
// standard 6x7-need-4 game and small variants useful for benchmarking
// optimal play (e.g. 4x4-need-3). Full 6x7 is far beyond a naive minimax
// and is left as a hook for richer search strategies.

use crate::game::{Game, Side};

#[derive(Clone, Copy, Debug)]
pub struct ConnectFourUndo {
    pub col: usize,
    pub row: usize,
    pub prev_winner: Option<Side>,
}

pub struct ConnectFour {
    rows: usize,
    cols: usize,
    need: usize,
    cells: Vec<char>,                // row-major, length rows*cols
    current: Side,
    winner: Option<Side>,
}

impl ConnectFour {
    pub fn new() -> Self {
        Self::with_size(6, 7, 4)
    }

    pub fn with_size(rows: usize, cols: usize, need: usize) -> Self {
        Self {
            rows, cols, need,
            cells: vec!['.'; rows * cols],
            current: Side::One,
            winner: None,
        }
    }

    fn idx(&self, r: usize, c: usize) -> usize { r * self.cols + c }
    fn at(&self, r: usize, c: usize) -> char { self.cells[self.idx(r, c)] }

    fn mark(s: Side) -> char {
        if s == Side::One { 'X' } else { 'O' }
    }

    // Count consecutive `m`s in direction (dr,dc) from (r,c), exclusive.
    fn run(&self, r: isize, c: isize, dr: isize, dc: isize, m: char) -> usize {
        let mut n = 0;
        for i in 1..self.need as isize {
            let rr = r + dr * i;
            let cc = c + dc * i;
            if rr < 0 || rr >= self.rows as isize || cc < 0 || cc >= self.cols as isize {
                break;
            }
            if self.at(rr as usize, cc as usize) != m { break; }
            n += 1;
        }
        n
    }

    fn check_win_at(&self, r: usize, c: usize) -> bool {
        let m = self.at(r, c);
        let dirs: [(isize, isize); 4] = [(0, 1), (1, 0), (1, 1), (1, -1)];
        for (dr, dc) in dirs {
            let a = self.run(r as isize, c as isize, dr, dc, m);
            let b = self.run(r as isize, c as isize, -dr, -dc, m);
            if 1 + a + b >= self.need { return true; }
        }
        false
    }
}

impl Default for ConnectFour {
    fn default() -> Self { Self::new() }
}

impl Game for ConnectFour {
    type Move = usize;
    type Undo = ConnectFourUndo;

    fn legal_moves(&self) -> Vec<Self::Move> {
        if self.winner.is_some() {
            return Vec::new();
        }
        (0..self.cols).filter(|&c| self.at(0, c) == '.').collect()
    }

    fn apply_move(&mut self, col: &Self::Move) -> Self::Undo {
        let col = *col;
        let prev_winner = self.winner;
        let mut row = self.rows - 1;
        while self.at(row, col) != '.' {
            if row == 0 { break; }
            row -= 1;
        }
        let idx = self.idx(row, col);
        self.cells[idx] = Self::mark(self.current);
        if self.check_win_at(row, col) {
            self.winner = Some(self.current);
        }
        self.current = self.current.other();
        ConnectFourUndo { col, row, prev_winner }
    }

    fn undo_move(&mut self, undo: Self::Undo) {
        let idx = self.idx(undo.row, undo.col);
        self.cells[idx] = '.';
        self.winner = undo.prev_winner;
        self.current = self.current.other();
    }

    fn current_side(&self) -> Side { self.current }
    fn is_over(&self) -> bool { self.winner.is_some() || self.legal_moves().is_empty() }
    fn winner(&self) -> Option<Side> { self.winner }

    fn state_key(&self) -> u64 {
        // rows*cols cells * 2 bits + 1 side bit. Caller must keep
        // rows*cols <= 31 for u64. Benchmark uses 4x4 = 16 cells = 33 bits.
        let mut k: u64 = 0;
        let mut i = 0;
        for r in 0..self.rows {
            for c in 0..self.cols {
                let v: u64 = match self.at(r, c) { '.' => 0, 'X' => 1, _ => 2 };
                k |= v << (2 * i);
                i += 1;
            }
        }
        k |= (if self.current == Side::One { 0u64 } else { 1u64 }) << (2 * i);
        k
    }

    fn render(&self) -> String {
        let mut s = String::new();
        for r in 0..self.rows {
            s.push('|');
            for c in 0..self.cols {
                s.push(self.at(r, c));
                s.push('|');
            }
            s.push('\n');
        }
        s.push(' ');
        for c in 0..self.cols {
            s.push_str(&format!("{} ", c));
        }
        s.push('\n');
        s
    }

    fn move_to_string(&self, col: &Self::Move) -> String {
        format!("col {}", col)
    }
}
